packetbeat/sniffer/sniffer.go

```diff
@@ -134,15 +140,124 @@ func New(testMode bool, filter string, factory WorkerFactory, interfaces config.
+
+		// Prime the channel.
+		current := s.device
+		device <- current
+
+		tick := time.NewTicker(s.config.PollDefaultRoute)
+		for {
+			select {
+			case <-tick.C:
+				logp.Debug("sniffer", "polling default route")
+				dev, err := resolveDeviceName(s.config.Device)
+				if err != nil {
+					logp.Warn("sniffer failed to poll default route device: %v", err)
+					continue
+				}
+				if dev != current {
+					logp.Info("sniffer changing default route device: %s -> %s", current, dev)
+					current = dev
+					s.state.Store(snifferInactive) // Mark current device as stale. ¯\_(ツ)_/¯
+					device <- current              // Pass the new device name.
+				}
+			case <-s.done:
+				logp.Info("closing default route poller")
+				close(device)
+				tick.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// sniffStatic performs the sniffing work on a single static interface.
+func (s *Sniffer) sniffStatic(device string) error {
+	handle, err := s.open(device)
 	if err != nil {
 		return fmt.Errorf("failed to start sniffer: %w", err)
 	}
 	defer handle.Close()
 
+	worker, err := s.factory(handle.LinkType())
+	if err != nil {
+		return err
+	}
+
+	return s.sniffHandle(handle, worker)
+}
+
+// sniffDynamic performs sniffing work on a stream of dynamic interfaces from
+// defaultRoute workers are retained between successive interfaces if they are
+// the same link type.
+func (s *Sniffer) sniffDynamic(defaultRoute <-chan string) error {
+	var (
+		last   layers.LinkType
+		worker Worker
+	)
+	for device := range defaultRoute {
+		var err error
+		last, worker, err = s.sniffOneDynamic(device, last, worker)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// sniffOneDynamic handles sniffing a single device that may change link type.
+// If the the link type associated with the device differs from the last link
+// type or worker is nil, a new worker is returned. The link type associated
+// with the device is returned.
+func (s *Sniffer) sniffOneDynamic(device string, last layers.LinkType, worker Worker) (layers.LinkType, Worker, error) {
+	handle, err := s.open(device)
+	if err != nil {
+		return last, worker, fmt.Errorf("failed to start sniffer: %w", err)
+	}
+	defer handle.Close()
+
+	linkType := handle.LinkType()
+	if worker == nil || linkType != last {
+		logp.Info("changing link type: %d -> %d", last, linkType)
+		worker, err = s.factory(linkType)
+		if err != nil {
+			return linkType, worker, err
+		}
+	}
+
+	err = s.sniffHandle(handle, worker)
+	return linkType, worker, err
+}
+
+// sniff performs the sniffing work and writing dump files if requested.
+func (s *Sniffer) sniffHandle(handle snifferHandle, worker Worker) error {
 	var w *pcapgo.Writer
 	if s.config.Dumpfile != "" {
-		f, err := os.Create(s.config.Dumpfile)
+		const timeSuffixFormat = "20060102150405"
+		filename := fmt.Sprintf("%s-%s.pcap", s.config.Dumpfile, time.Now().Format(timeSuffixFormat))
+		logp.Info("creating new dump file %s", filename)
+		f, err := os.Create(filename)
```

Comment:
This will result in file clobbering if more than one sniffer is run by the config manager. This is not worse than what we currently have, but worth being aware of.